### PR TITLE
New DailyDialog Teacher to exclude __SILENCE__ tokens 

### PR DESCRIPTION
**Pull request by ricsinaruto**
Made the modifications according to #1659 
Basically added a modified dailydialog version where there is no START_ENTRY.

I still tried to keep the two speaker setup however, to have as many examples as possible. Basically, when the agent should start the dialog the first utterance is skipped, and the second utterance is the input. 

However, when the dialog consists of only 2 turns we can't shift the dialog, so I just use the same dialog regardless of who should start the dialog. Because all episodes are duplicated and there has to be at least 1 entry for each episode, I don't see another way to solve this, but I don't think it should be a big problem that the few 2-turn dialogs that are in the dataset are duplicated.

The naming of the datasets might not be the best, I'm open for suggestions :)

Tested with the display_data script, seems to work as expected (22236 episodes with a total of 87170 examples)
image

I tried to run this command from the contributing guidelines, but it gave me an error:
```
python setup.py test -s tests.suites.datatests
```
Error:
```
running test
Searching for markdown<3.0
Best match: Markdown 2.6.11
Processing Markdown-2.6.11-py3.6.egg

Using /mnt/permanent/home/csaky/miniconda3/envs/python-3.6-gpu/lib/python3.6/site-packages/Markdown-2.6.11-py3.6.egg
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
datasets (unittest.loader._FailedTest) ... ERROR

======================================================================
ERROR: datasets (unittest.loader._FailedTest)
----------------------------------------------------------------------
AttributeError: module 'tests.suites' has no attribute 'datasets'

----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
```

### CODE REVIEW

**Comment by ricsinaruto** [comment]
That one unit test that failed seems totally unrelated to my modifications

**Comment by stephenroller** [comment]
Are you sure you ran `python setup.py test -s tests.suites.datatests`? It looks like you ran `python setup.py test -s tests.suites.datasets`, but I'm not discounting the possibility of outdated docs or a bug.

**Comment by ricsinaruto** [comment]
> Are you sure you ran `python setup.py test -s tests.suites.datatests`? It looks like you ran `python setup.py test -s tests.suites.datasets`, but I'm not discounting the possibility of outdated docs or a bug.

My bad!
Here is the output:
```
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
test_verify_data (test_new_tasks.TestNewTasks) ... ok

----------------------------------------------------------------------
Ran 1 test in 46.728s

OK
```

**Comment by ricsinaruto** [comment]
> I'm not sure the number of examples is correct. It's set to the total number of utterances before, but now one utterance per conversation needs to be excluded, no?

You are right, the number of examples should be less, I have no idea why the display_data script returns this number.

**Comment by stephenroller** [comment]
https://github.com/facebookresearch/ParlAI/blob/da41e279af0ba3c63842f12318ec686981acce38/parlai/tasks/dailydialog/agents.py#L47-L57

**Comment by ricsinaruto** [comment]
Okay the reason is, that it is pre-computed by a function (num_examples) in the original dataset, so I have to change this.

**Comment by ricsinaruto** [comment]
Added a commit to fix it, the correct number of examples is 76052.

File: parlai/tasks/dailydialog/agents.py [hunk 1]
```diff
@@ -92,6 +92,13 @@ def share(self):
 
 class NoSilenceTeacher(Convai2Teacher):
     # Same as the default teacher, but it doesn't contain __SILENCE__ entries.
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        # Calculate the correct number of examples.
+        dialogs = [d['dialogue'] for d in self.data]
+        self.num_exs -= len([0 for d in dialogs if len(d) >= 2])
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: Can we use an explicit sum here?

I think it's `self.num_exs = sum(len(d['dialogue']) - 1 for d in self.data))`?

Comment (ricsinaruto) [reply parlai/tasks/dailydialog/agents.py]: Done!

File: parlai/tasks/dailydialog/agents.py [hunk 2]
```diff
@@ -88,3 +88,43 @@ def share(self):
         shared = super().share()
         shared['data'] = self.data
         return shared
+
+
+class NoSilenceTeacher(Convai2Teacher):
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: What do you think about `NoStart`? `NoSilence` is fine but I agree we could have a slightly better name

File: parlai/tasks/dailydialog/agents.py [hunk 3]
```diff
@@ -88,3 +88,43 @@ def share(self):
...
+    # Same as the default teacher, but it doesn't contain __SILENCE__ entries.
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: Make this into a docstring instead please.

File: parlai/tasks/task_list.py [hunk 4]
```diff
@@ -889,6 +889,20 @@
             "for more information."
         )
     },
+    {
```
Comment (stephenroller) [inline parlai/tasks/task_list.py]: Let's not add this to task list.

File: parlai/tasks/dailydialog/agents.py [hunk 5]
```diff
@@ -88,3 +88,43 @@ def share(self):
...
+
+    def get(self, episode_idx, entry_idx=0):
+        # Sometimes we're speaker 1 and sometimes we're speaker 2
+        speaker_id = episode_idx % 2
+        full_eps = self.data[episode_idx // 2]
+        entries = full_eps['dialogue']
+
+        # If we are first speaker, then the first utterance is skipped.
+        # If dialog is only 2 turns, then we can't shift it.
+        if len(entries) > 2:
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: Honestly it feels like we should be able to do this without special casing using correct math. If a conversation only has 2 turns, then we shouldn't have it included in num_episodes. This makes all of this much more nuanced :(

Comment (ricsinaruto) [reply parlai/tasks/dailydialog/agents.py]: I will think about it, it might be easier than I originally thought

**Comment by ricsinaruto** [comment]
I think it should be okay now, I just store all episodes separately so the 2-turn dialogs are only stored once.
21753 episodes with a total of 76052 examples.

File: parlai/tasks/dailydialog/agents.py [hunk 6]
```diff
@@ -88,3 +88,44 @@ def share(self):
...
+        self.all_eps = self.data + [d for d in self.data if len(d['dialogue']) > 2]
+        self.num_eps = len(self.all_eps)
+
+    def get(self, episode_idx, entry_idx=0):
+        full_eps = self.all_eps[episode_idx]
+        entries = full_eps['dialogue']
+
+        # Sometimes we're speaker 1 and sometimes we're speaker 2.
+        # We can't be speaker 1 if dialog has only 2 turns.
+        speaker_id = episode_idx % 2 if len(entries) > 2 else 0
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: I think this should now be `speaker_id = int(episode_idx >= len(self.data))`, right?

File: parlai/tasks/dailydialog/agents.py [hunk 7]
```diff
@@ -88,3 +88,44 @@ def share(self):
...
+        If we are the first speaker, then the first utterance is skipped.
+    """
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        # Calculate the correct number of examples.
+        self.num_exs = sum(len(d['dialogue']) - 1 for d in self.data)
+
+        # Store all episodes in self.data
+        self.all_eps = self.data + [d for d in self.data if len(d['dialogue']) > 2]
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: good

File: parlai/tasks/dailydialog/agents.py [hunk 8]
```diff
@@ -88,3 +88,44 @@ def share(self):
         shared = super().share()
         shared['data'] = self.data
         return shared
+
+
+class NoStartTeacher(Convai2Teacher):
+    """ Same as default teacher, but it doesn't contain __SILENCE__ entries.
```
Comment (stephenroller) [inline parlai/tasks/dailydialog/agents.py]: indenting
```python
"""
Same as default teacher, but it doesn't contain __SILENCE__ entries.
If we are the first speaker, then the first utterance is skipped.
"""
```

File: parlai/tasks/dailydialog/agents.py (hunk 6 above)
Comment (ricsinaruto) [reply parlai/tasks/dailydialog/agents.py]: Yes, that's actually a lot better.

**Comment by stephenroller** [comment]
Cool, one more run of `python setup.py test -s tests.suites.datatests` and lgtm.

**Comment by ricsinaruto** [comment]
> Cool, one more run of `python setup.py test -s tests.suites.datatests` and lgtm.

```
running test
Searching for markdown<3.0
Best match: Markdown 2.6.11
Processing Markdown-2.6.11-py3.6.egg

Using /mnt/permanent/home/csaky/miniconda3/envs/python-3.6-gpu/lib/python3.6/site-packages/Markdown-2.6.11-py3.6.egg
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
test_verify_data (test_new_tasks.TestNewTasks) ... ok

----------------------------------------------------------------------
Ran 1 test in 80.768s

OK
```

**Pull request merged**